packages/core/src/manifest.ts

```diff
@@ -38,27 +40,82 @@ function defaultManifest(): ManifestData {
   };
 }
 
-const manifestDir = '.openzeppelin';
+const MANIFEST_DEFAULT_DIR = '.openzeppelin';
+const MANIFEST_TEMP_DIR = 'openzeppelin';
+
+async function getDevelopmentInstanceId(provider: EthereumProvider, chainId: number): Promise<string | undefined> {
+  let hardhatMetadata;
+
+  try {
+    hardhatMetadata = await getHardhatMetadata(provider);
+  } catch (e: unknown) {
+    if (e instanceof Error && (e.message.includes('Method') || e.message.includes('method'))) {
+      // method not found
+      return undefined;
+    } else {
+      throw e;
+    }
+  }
+
+  if (hardhatMetadata.chainId !== chainId) {
+    throw new Error(
+      `Broken invariant: Hardhat metadata's chainId ${hardhatMetadata.chainId} does not match eth_chainId ${chainId}`,
+    );
+  }
+
+  return hardhatMetadata.instanceId;
+}
+
+function getSuffix(chainId: number, devInstanceId?: string) {
+  if (devInstanceId !== undefined) {
+    return `${chainId}-${devInstanceId}`;
+  } else {
+    return chainId;
+  }
+}
+
+async function makeAndGetTempDir() {
+  const tempDir = path.join(os.tmpdir(), MANIFEST_TEMP_DIR);
+  await fs.mkdir(tempDir, { recursive: true });
+  return tempDir;
+}
 
 export class Manifest {
   readonly chainId: number;
+  readonly devInstanceId: string | undefined;
   readonly file: string;
+  readonly manifestDir: string;
   private readonly fallbackFile: string;
 
   private locked = false;
 
   static async forNetwork(provider: EthereumProvider): Promise<Manifest> {
-    return new Manifest(await getChainId(provider));
+    const chainId = await getChainId(provider);
+    const devInstanceId = await getDevelopmentInstanceId(provider, chainId);
+    const dir = devInstanceId !== undefined ? await makeAndGetTempDir() : undefined;
+
+    return new Manifest(chainId, devInstanceId, dir);
   }
 
-  constructor(chainId: number) {
+  constructor(chainId: number, devInstanceId?: string, dir?: string) {
     this.chainId = chainId;
+    this.devInstanceId = devInstanceId;
+    this.manifestDir = dir ?? MANIFEST_DEFAULT_DIR;
+    debug('manifest dir:', this.manifestDir);
+
+    const fallbackName = `unknown-${getSuffix(chainId, devInstanceId)}`;
```

Comment:
I'd like to start using a name other than `unknown` for these files. One option would be to use `dev-...`, but if we know these are Hardhat we could go with `hardhat-...`.